src/moonlink/src/error.rs

```diff
@@ -1,109 +1,280 @@
 #[derive(Clone, Debug, Error)]
 pub enum Error {
-    #[error("Arrow error: {source}")]
-    Arrow { source: Arc<ArrowError> },
+    #[error("{0}")]
+    Arrow(ErrorStruct),
 
-    #[error("IO error: {source}")]
-    Io { source: Arc<io::Error> },
+    #[error("{0}")]
+    Io(ErrorStruct),
 
-    #[error("Parquet error: {source}")]
-    Parquet { source: Arc<ParquetError> },
+    #[error("{0}")]
+    Parquet(ErrorStruct),
 
     #[error("Transaction {0} not found")]
     TransactionNotFound(u32),
 
-    #[error("Watch channel receiver error: {source}")]
-    WatchChannelRecvError {
-        #[source]
-        source: watch::error::RecvError,
-    },
+    #[error("{0}")]
+    WatchChannelRecvError(ErrorStruct),
 
     #[error("Tokio join error: {0}. This typically occurs when a spawned task fails to complete successfully. Check the task's execution or panic status for more details.")]
     TokioJoinError(String),
 
-    #[error("Iceberg error: {source}")]
-    IcebergError { source: Arc<IcebergError> },
+    #[error("{0}")]
+    IcebergError(ErrorStruct),
 
     #[error("Iceberg error: {0}")]
     IcebergMessage(String),
 
-    #[error("OpenDAL error: {source}")]
-    OpenDal { source: Arc<opendal::Error> },
+    #[error("{0}")]
+    OpenDal(ErrorStruct),
 
     #[error("UTF-8 conversion error: {0}")]
     Utf8(#[from] std::string::FromUtf8Error),
 
-    #[error("Join error: {source}")]
-    JoinError { source: Arc<tokio::task::JoinError> },
+    #[error("{0}")]
+    JoinError(ErrorStruct),
 
-    #[error("JSON serialization/deserialization error: {source}")]
-    Json { source: Arc<serde_json::Error> },
+    #[error("{0}")]
+    Json(ErrorStruct),
 }
 
 pub type Result<T> = result::Result<T, Error>;
 
+impl From<watch::error::RecvError> for Error {
+    fn from(source: watch::error::RecvError) -> Self {
+        Error::WatchChannelRecvError(ErrorStruct {
+            message: format!("Watch channel receiver error: {source}"),
+            status: ErrorStatus::Permanent,
+            source: Some(Arc::new(source.into())),
+        })
+    }
+}
+
 impl From<ArrowError> for Error {
     fn from(source: ArrowError) -> Self {
-        Error::Arrow {
-            source: Arc::new(source),
-        }
+        let status = match source {
+            ArrowError::MemoryError(_) | ArrowError::IoError(_, _) => ErrorStatus::Temporary,
+
+            // All other errors are regard as permanent
+            _ => ErrorStatus::Permanent,
+        };
+
+        Error::Arrow(ErrorStruct {
+            message: format!("Arrow error: {source}"),
+            status,
+            source: Some(Arc::new(source.into())),
+        })
     }
 }
 
 impl From<IcebergError> for Error {
     fn from(source: IcebergError) -> Self {
-        Error::IcebergError {
-            source: Arc::new(source),
-        }
+        Error::IcebergError(ErrorStruct {
+            message: format!("Iceberg error: {source}"),
+            status: ErrorStatus::Permanent,
```

Comment:
- Commit conflict is a retryable error: https://github.com/apache/iceberg-rust/blob/299cfedb7abc3d130268c519f8afe14c0e6728c2/crates/iceberg/src/error.rs#L65
- (no strong preference here) `Unexpected` error could be retryable also: https://github.com/apache/iceberg-rust/blob/299cfedb7abc3d130268c519f8afe14c0e6728c2/crates/iceberg/src/error.rs#L37C5-L37C15, it's used extensively inside of moonlink to represent IO errors; eg: https://github.com/Mooncake-Labs/moonlink/blob/c91807cedf2a2e008c919c6a4784baed776449d7/src/moonlink/src/storage/iceberg/file_catalog.rs#L246